doc/feature/compartmentReport.md

```diff
@@ -0,0 +1,467 @@
+
+  For getOffsets():
+
+       /** Get the current mapping of each section of each neuron in each
+        *  simulation frame buffer.
+        *
+        * For instance, getOffsets()[1][15] retrieves the lookup index for the
+        * frame buffer for section 15 of neuron with index 1. The neuron index
+        * is derived from the order in the GID set provided by updateMapping().
+        *
+        * After construction or updateMapping is called, the result is empty
+        * until a frame is read.
+        *
+        * For requested ceels that are not part of the report its section
+        * offset list will be empty.
+        *
+        * @return the offset for each section for each neuron
+        */
+
+  For getCompartmentCounts():
+
+       /** Get the number of compartments for each section of each neuron
+        *  provided by the GID set via updateMapping().
+        *
+        * The neuron index is derived from the order in the GID set provided by
+        * updateMapping(). For requested GIDs that are not part of the report
+        * its compartment count list will be empty.
+        *
+        * After construction or updateMapping is called, the result is empty
+        * until a frame is read.
+        *
+        * @return the compartment counts for each section for each neuron
+        * @version 1.0
+        */
+       const CompartmentCounts& getCompartmentCounts() const;
+
+ getNumCompartments():
+
+       /** Get the number of compartments for the given neuron.
+        *
+        * The neuron index is derived from the order in the GID set provided by
+        * updateMapping().
+        *
+        * @param index neuron index per current GID set
+        * @return number of compartments for the given neuron or 0 if the
+        *         report has just been created or updateMapping is called, and
+        *         not frame has been read yet. Also 0 for neurons not part of
+        *         the report at the source.
+        */
+
+  Note that the new functions do not return valid data until the first frame
+  is read.
+
+- Replace getEndTime() documentation by:
+
+       /** @return the most up to date end time of the report.
+        *
+        * This value may change upon the completion of readNextFrame, seek,
+        * or writeHeader.
+        */
+       float getEndTime() const;
+
+- Add the following preconditions to writeHeader:
+
+        * - The report was open in write mode.
+        * - The header has not already been written.
+
+## Implementation
+
+This section will cover some implementation details related to the stream
+case as the file based IO is straight forward.
+
+The most two important decisions taken are:
+- The granularity of the messages sent won't be the whole frame but a cell.
+- The mapping will be always included as part of the frame.
+
+The rationale for these decisions follows.
+
+The current API for writing is already cell oriented, while this is convenient
```

Comment:
It came to my mind, but it hasn't been discussed. My feeling is that some buffering will still be needed and the synchronization at frame boundaries may be more difficult if the API doesn't have a function to read a full frame atomically. If you want to have a whiteboard discussion I'm open to it, otherwise feel free to provide a more concrete proposal.
Something that was proposed and probably is a good idea is to make all mapping functions get a GID as input parameter. That frees the user from having to do the GID->index translation, which has already caused confusion in the past.